In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split


from sklearn.datasets import load_iris

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,precision_score,recall_score

from sklearn.metrics import confusion_matrix,classification_report

from sklearn.model_selection import GridSearchCV

In [ ]:
df=load_iris(as_frame=True).frame

df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,2
146,6.3,2.5,5.0,1.9,2
147,6.5,3.0,5.2,2.0,2
148,6.2,3.4,5.4,2.3,2


In [7]:
x = df.drop(['target'],axis=1)
y = df['target']
x_train , x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42,stratify=y)

In [10]:
model=DecisionTreeClassifier()
model.fit(x_train,y_train)
y_pred= model.predict(x_test)
y_pred

array([0, 2, 1, 1, 0, 1, 0, 0, 2, 1, 2, 2, 2, 1, 0, 0, 0, 1, 1, 2, 0, 2,
       1, 1, 2, 2, 1, 0, 2, 0])

In [11]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       0.90      0.90      0.90        10
           2       0.90      0.90      0.90        10

    accuracy                           0.93        30
   macro avg       0.93      0.93      0.93        30
weighted avg       0.93      0.93      0.93        30



In [ ]:
grid ={
    "criterion":['gini','entropy'],
    'max_depth':[2,3,4,5,6,7,8,None],
}

grid_search=GridSearchCV(
    estimator=model,
    param_grid=grid,
    cv=5
    )
grid_search.fit(x_train,y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 3, 4, 5, 6, 7, 8, None]})

In [ ]:
y_pred = grid_search.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       0.90      0.90      0.90        10
           2       0.90      0.90      0.90        10

    accuracy                           0.93        30
   macro avg       0.93      0.93      0.93        30
weighted avg       0.93      0.93      0.93        30



#short note :

**Gini Impurity** মানে হলো—

> **একটি Node-এর Data কতটা মিশ্রিত (Impure) আছে, তার পরিমাপ।**

অথবা আরও সহজভাবে,

> **একটি Node থেকে Random একটি Sample নিলে সেটিকে ভুল Classify হওয়ার সম্ভাবনা কত?**

---

## উদাহরণ ১: Pure Node

ধরো একটি Node-এ আছে

```text
10 টি ফুল

Setosa = 10
Versicolor = 0
Virginica = 0
```

সবগুলো একই Class।

এখানে

**Gini = 0**

মানে **একদম Pure**।

---

## উদাহরণ ২: Mixed Node

ধরো

```text
10 টি ফুল

Setosa = 5
Versicolor = 5
```

এখানে দুইটা Class মিশে আছে।

**Gini = 0.5**

মানে Node Impure।

---

## উদাহরণ ৩

ধরো

```text
10 টি ফুল

Setosa = 4
Versicolor = 3
Virginica = 3
```

এখানে আরও বেশি Mixing।

Gini আরও বেশি হবে।

---

## Formula

[
\text{Gini} = 1 - \sum p_i^2
]

এখানে,

* (p_i) = প্রতিটি Class-এর Probability

---

### ছোট Example

ধরো

```text
Setosa = 8
Versicolor = 2
```

Probability

```text
Setosa = 8/10 = 0.8

Versicolor = 2/10 = 0.2
```

তাহলে

[
\text{Gini}
===========

1-(0.8^2+0.2^2)
]

[
=1-(0.64+0.04)
]

[
=0.32
]

অর্থাৎ **Gini = 0.32**।

---

## Decision Tree এটা কীভাবে ব্যবহার করে?

ধরো দুটি Feature আছে।

### Split 1

```text
Gini = 0.42
```

### Split 2

```text
Gini = 0.12
```

Decision Tree বলবে,

👉 **Split 2 ভালো**, কারণ Gini কম মানে Node বেশি Pure।

---

## সহজে মনে রাখো

* **Gini = 0** → একদম Pure (সব একই Class) ✅
* **Gini বেশি** → Class গুলো বেশি মিশে আছে ❌

---

### ১ লাইনে Viva Answer

**"Gini Impurity is a measure of how mixed the classes are in a node. A lower Gini value indicates a purer node, so Decision Tree chooses the split that minimizes Gini Impurity."**


# Entropy note

অবশ্যই। **Entropy**-কে Gini-এর মতোই সহজভাবে বুঝি।

---

# Entropy মানে কী?

**Entropy** হলো—

> **একটি Node-এর Data কতটা বিশৃঙ্খল (Disordered) বা Mixed আছে, তার পরিমাপ।**

অর্থাৎ,

* **Entropy কম = Data পরিষ্কার (Pure)**
* **Entropy বেশি = Data মিশ্রিত (Impure)**

---

## উদাহরণ ১: Pure Node

ধরো একটি Node-এ আছে

```text
10 টি ফুল

Setosa = 10
Versicolor = 0
Virginica = 0
```

সব একই Class।

তাহলে

```text
Entropy = 0
```

মানে **কোনো Confusion নেই।**

---

## উদাহরণ ২: Mixed Node

ধরো

```text
10 টি ফুল

Setosa = 5
Versicolor = 5
```

এখন দুইটা Class সমান।

তাহলে

```text
Entropy = 1
```

মানে **Confusion সবচেয়ে বেশি।**

---

## উদাহরণ ৩

ধরো

```text
10 টি ফুল

Setosa = 8
Versicolor = 2
```

এখানে

Data পুরোপুরি Mixed না।

তাই

```text
Entropy ≈ 0.72
```

অর্থাৎ কিছুটা Confusion আছে।

---

# Formula

[
Entropy=-\sum p_i\log_2(p_i)
]

এখানে

* (p_i) = প্রতিটি Class-এর Probability

---

### Example

ধরো

```text
Setosa = 8
Versicolor = 2
```

Probability

```text
Setosa = 0.8

Versicolor = 0.2
```

তাহলে

[
Entropy
=======

-(0.8\log_2 0.8+0.2\log_2 0.2)
]

[
≈0.72
]

---

# Decision Tree কীভাবে ব্যবহার করে?

ধরো দুটি Split পাওয়া গেল।

### Split 1

```text
Entropy = 0.60
```

### Split 2

```text
Entropy = 0.15
```

Decision Tree বলবে

👉 **Split 2 ভালো।**

কারণ

**Entropy কম = Node বেশি Pure।**

---

# তাহলে Information Gain কোথায় আসে?

Decision Tree আগে দেখে

Split করার আগে Entropy কত।

ধরো

```text
Before Split

Entropy = 1
```

Split করার পরে

```text
After Split

Entropy = 0.2
```

তাহলে

```text
Information Gain

= 1 - 0.2

= 0.8
```

অর্থাৎ

**Split করার ফলে Confusion অনেক কমে গেছে।**

তাই Decision Tree এই Split-টাই বেছে নেবে।

---



## খুব সহজ ট্রিক

ধরো একটা ক্লাসে ১০ জন ছাত্র আছে।

### সবাই একই সেকশনের

```text
A A A A A A A A A A
```

👉 কোনো Confusion নেই।

**Entropy = 0**

---

### অর্ধেক A, অর্ধেক B

```text
A A A A A
B B B B B
```

👉 কে কোন সেকশনের, বুঝতে Confusion বেশি।

**Entropy = সবচেয়ে বেশি (১)**

---

### Viva-তে ১ লাইনের উত্তর

> **Entropy is a measure of the uncertainty or randomness in a node. Lower entropy means the node is purer, while higher entropy means the classes are more mixed.**


# Gini vs Entropy

| Gini                                | Entropy                                                               |
| ----------------------------------- | --------------------------------------------------------------------- |
| Data কতটা Mixed তা মাপে             | Data কতটা Confused তা মাপে                                            |
| Gini কম হলে ভালো                    | Entropy কম হলে ভালো                                                   |
| Gini ব্যবহার করে Split নির্বাচন করে | Information Gain (Entropy কমার পরিমাণ) ব্যবহার করে Split নির্বাচন করে |

---

# CODE IMPLEMNETATION

অবশ্যই। এই কোডটা মূলত **Decision Tree Classification** করে, তারপর **GridSearchCV** ব্যবহার করে সবচেয়ে ভালো hyperparameter খুঁজে বের করে। একদম লাইন বাই লাইন বুঝাই।

---

# ১. প্রয়োজনীয় লাইব্রেরি Import

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
```

এগুলো সাধারণ Data Analysis এর জন্য।

* `numpy` → Mathematical calculation
* `pandas` → DataFrame নিয়ে কাজ
* `matplotlib` → Graph আঁকা
* `seaborn` → সুন্দর graph

---

```python
from sklearn.model_selection import train_test_split
```

Dataset কে Train এবং Test এ ভাগ করার জন্য।

---

```python
from sklearn.datasets import load_iris
```

Scikit-learn এর built-in Iris Dataset লোড করার জন্য।

---

```python
from sklearn.tree import DecisionTreeClassifier
```

Decision Tree Model ব্যবহার করার জন্য।

---

```python
from sklearn.metrics import accuracy_score, precision_score, recall_score
```

Model কত ভালো করেছে সেটা মাপার জন্য।

---

```python
from sklearn.metrics import confusion_matrix, classification_report
```

Classification Report এবং Confusion Matrix দেখানোর জন্য।

---

```python
from sklearn.model_selection import GridSearchCV
```

Model এর সবচেয়ে ভালো parameter বের করার জন্য।

---

# ২. Dataset Load করা

```python
df = load_iris(as_frame=True).frame
```

`load_iris()` Iris Dataset নিয়ে আসে।

`as_frame=True`

মানে Dataset কে DataFrame হিসেবে দিবে।

তারপর

```python
.frame
```

লিখলে পুরো DataFrame পাওয়া যায়।

---

ধরি Dataset এমনঃ

| sepal length | sepal width | petal length | petal width | target |
| ------------ | ----------- | ------------ | ----------- | ------ |
| 5.1          | 3.5         | 1.4          | 0.2         | 0      |
| 4.9          | 3.0         | 1.4          | 0.2         | 0      |

---

```python
df
```

Dataset দেখাবে।

---

# ৩. Feature এবং Target আলাদা করা

```python
X = df.drop(['target'], axis=1)
```

`drop()` মানে target column বাদ দাও।

তাহলে

X =

```
sepal length
sepal width
petal length
petal width
```

---

```python
y = df['target']
```

Target আলাদা করলাম।

```
0
1
2
0
1
...
```

---

# ৪. Train Test Split

```python
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)
```

এখানে

```
test_size=0.2
```

মানে

২০% Test

৮০% Train

---

```
random_state=42
```

যাতে প্রতিবার একইভাবে Dataset ভাগ হয়।

---

```
stratify=y
```

এটা খুব গুরুত্বপূর্ণ।

Iris এ ৩টা Class আছে।

```
0
1
2
```

Stratify দিলে তিনটা Class-ই Train এবং Test এ সমান অনুপাতে থাকবে।

---

# ৫. Model তৈরি

```python
model = DecisionTreeClassifier()
```

Decision Tree Object বানালাম।

এখনও Train হয়নি।

---

# ৬. Model Train

```python
model.fit(X_train, y_train)
```

এখানে Model শেখে।

Input

```
X_train
```

Output

```
y_train
```

Decision Tree এখন Rules তৈরি করবে।

যেমন

```
Petal Length <=2 ?

Yes -> Class 0

No -> নিচে যাও

Petal Width <=1.8 ?

Yes -> Class 1

No -> Class 2
```

---

# ৭. Prediction

```python
y_pred = model.predict(X_test)
```

এখন Model Test Data দেখে Prediction করছে।

ধরি

Actual

```
0
1
2
```

Prediction

```
0
1
2
```

---

# ৮. Report

```python
print(classification_report(y_test, y_pred))
```

Output হতে পারে

```
precision

recall

f1-score

support
```

---

ধরি

```
precision = 0.96
```

মানে

Model যতবার Class 1 বলেছে

তার 96% সত্যি ছিল।

---

```
recall = 0.94
```

মানে

Actual Class 1 যত ছিল

তার 94% ধরতে পেরেছে।

---

```
f1-score
```

Precision এবং Recall এর Balance।

---

# এখন Hyperparameter Tuning

এখান থেকেই GridSearchCV শুরু।

---

## Grid বানানো

```python
grid = {
    "criterion": ['gini', 'entropy'],
    "max_depth": [1,2,3,4,5,6,7,8,None]
}
```

এখানে Model এর বিভিন্ন Parameter দেওয়া হয়েছে।

---

### criterion

Decision Tree কীভাবে Split করবে।

```
gini
```

বা

```
entropy
```

দুটোই Split করার Method।

---

### max_depth

Tree কত গভীর হবে।

```
1
```

```
2
```

```
3
```

....

```
None
```

মানে Unlimited।

---

Grid এর Combination হবে

| criterion | max_depth |
| --------- | --------- |
| gini      | 1         |
| gini      | 2         |
| gini      | 3         |
| ...       | ...       |
| entropy   | 1         |
| entropy   | 2         |
| entropy   | 3         |

মোট

```
2 × 9 = 18
```

Combination।

---

# GridSearchCV

```python
grid_search = GridSearchCV(
    estimator=model,
    param_grid=grid,
    cv=5
)
```

এখানে

```
estimator=model
```

মানে কোন Model Optimize করবে।

---

```
param_grid=grid
```

মানে কোন কোন Parameter Try করবে।

---

```
cv=5
```

মানে

**5-Fold Cross Validation**

---

ধরি

Train Data = 120

এটাকে ৫ ভাগ করল।

```
Part1
Part2
Part3
Part4
Part5
```

---

Iteration 1

```
Train

2 3 4 5

Test

1
```

Iteration 2

```
Train

1 3 4 5

Test

2
```

এভাবে ৫ বার।

তারপর Average Accuracy বের করবে।

---

এভাবে সব

১৮টা Combination

Try করবে।

---

ধরি Result

| criterion | max_depth | Accuracy |
| --------- | --------- | -------- |
| gini      | 2         | 0.93     |
| gini      | 3         | 0.95     |
| entropy   | 3         | 0.98     |
| entropy   | 4         | 0.97     |

সবচেয়ে ভালো হলো

```
criterion = entropy

max_depth = 3
```

---

# Fit

```python
grid_search.fit(X_train, y_train)
```

এখানে GridSearchCV

সব Combination Try করে।

Best Model Save করে।

---

# Prediction

```python
y_pred = grid_search.predict(X_test)
```

এবার Normal Model না।

Best Model দিয়ে Prediction করছে।

---

# আবার Report

```python
print(classification_report(y_test, y_pred))
```

এখানে আগের তুলনায় Accuracy বা অন্যান্য Metric ভালো হতে পারে, কারণ GridSearchCV সবচেয়ে ভালো Parameter বেছে নিয়েছে।

---

## পুরো Flow এক নজরে

```
Load Dataset
        │
        ▼
Feature & Target আলাদা
        │
        ▼
Train-Test Split
        │
        ▼
Decision Tree Model তৈরি
        │
        ▼
Train (fit)
        │
        ▼
Predict
        │
        ▼
Classification Report
        │
        ▼
GridSearchCV
        │
        ▼
সব Parameter Combination Try
        │
        ▼
Best Parameter নির্বাচন
        │
        ▼
Best Model Train
        │
        ▼
Predict
        │
        ▼
Final Classification Report
```

**একটা গুরুত্বপূর্ণ বিষয়:** GridSearchCV চালানোর পর চাইলে সবচেয়ে ভালো parameter দেখতে পারো—

```python
print(grid_search.best_params_)
```

এবং সবচেয়ে ভালো Cross Validation Score দেখতে পারো—

```python
print(grid_search.best_score_)
```

এগুলো প্রায়ই Viva বা Lab-এ জিজ্ঞেস করা হয়।
